sync3/handler/handler.go

```diff
@@ -818,7 +834,46 @@ func (h *SyncLiveHandler) OnInvalidateRoom(p *pubsub.V2InvalidateRoom) {
 	ctx, task := internal.StartTask(context.Background(), "OnInvalidateRoom")
 	defer task.End()
 
-	h.Dispatcher.OnInvalidateRoom(ctx, p.RoomID)
+	// 1. Reload the global cache.
+	h.GlobalCache.OnInvalidateRoom(ctx, p.RoomID)
+
+	// Work out who is affected.
+	joins, invites, leaves, err := h.Storage.FetchMemberships(p.RoomID)
+	involvedUsers := make([]string, 0, len(joins)+len(invites)+len(leaves))
+	involvedUsers = append(involvedUsers, joins...)
+	involvedUsers = append(involvedUsers, invites...)
+	involvedUsers = append(involvedUsers, leaves...)
+
+	// 2. Reload the joined-room tracker.
+	if err != nil {
+		hub := internal.GetSentryHubFromContextOrDefault(ctx)
+		hub.WithScope(func(scope *sentry.Scope) {
+			scope.SetContext(internal.SentryCtxKey, map[string]any{
+				"room_id": p.RoomID,
+			})
+			hub.CaptureException(err)
+		})
+		logger.Err(err).
+			Str("room_id", p.RoomID).
+			Msg("Failed to fetch members after cache invalidation")
+	}
+
+	h.Dispatcher.OnInvalidateRoom(p.RoomID, joins, invites)
+
+	// 3. Destroy involved users' caches.
+	for _, userID := range involvedUsers {
```

Comment:
`involvedUsers` is potentially very large (think Matrix HQ). It would be better to cut this down to the connected users first, then continue. Otherwise we will spin a lot of cycles/locks on users not even on the upstream HS...

Reply:
> It would be better to cut this down to the connected users first, then continue.

Does this run the risk of introducing a race?

Reply:
PTAL at 78b1e59 and 4011e38.

Reply:
And 9aa8f55 :facepalm: